### ISSUE TITLE
Architectural Ideas

### ISSUE THREAD

**Issue by shmax** [erayd/json-schema-info#2]
Interesting idea, thanks for putting this together. I have two discussion points to kick things off with:

1. Abstract rules out such that they are language-agnostic. This might mean that they have to go in a separate repo, not sure, but the idea is that the rules are encoded once in a single place, and then any language can consume them, not just PHP. The format could be anything, really, but obvious suggestions are JSON, YAML, XML, and, well, I'll stop there. Perhaps it could live in the official json-schema-spec repo?

2. Separate the draft versions. Getting back to PHP and putting *#* 1 aside for the moment, the suggestion is to rearrange things such that lower schemas have no knowledge of the higher schemas. A few obvious ideas are to use a class hierarchy, or a system of config files that would be layered together. We'll have to sort out the details along the way, but I think that with either system all rules could be queried with any level of draft, and all rules would be implicitly off.

**Comment by erayd** [comment]
1) Great idea. I like the idea of JSON as a definition format, because everybody and his dog has a parser for it and we're already dealing with JSON anyway. It doesn't have comments though, which means we're going to need to add an extra attribute for the spec reference comments. Does that sound OK?

 2) This is already the case. None of the draft versions have any knowledge of the others (although they do all build off a common base). I have no issue with further separation however.

**Comment by erayd** [comment]
One other thought, that I feel is very important - whatever mechanism we use, needs to ensure that *all* feature checks are *always* available, regardless of which version of the standard is loaded.

**Comment by shmax** [comment]
1. I dunno. I think comments would be useful, but we can kick it around some more. I believe that yaml support is probably about as ubiquitous as JSON support, at least in the major languages. PHP definitely supports it. What do you mean by "extra attribute for the spec reference comments"?

2. I must be missing something. You currently have a single file with everything jammed into it (`setDraft03`, `setDraft04`, etc). The same piece of code knows about all the different specs, which means it will grow *ad infinitum* until the end of time.

> needs to ensure that all feature checks are always available, regardless of which version of the standard is loaded

Yes, I said that: "with either system all rules could be queried with any level of draft"

**Comment by erayd** [comment]
> I believe that yaml support is probably about as ubiquitous as JSON support, at least in the major languages. PHP definitely supports it.

PHP's YAML support is either a PECL extension, or a separate library - it doesn't have native YAML support that I'm aware of. I'd prefer to avoid it for that reason.

> What do you mean by "extra attribute for the spec reference comments"?

I mean that the spec reference comments need a home. Because JSON does not support comments, if we use JSON they would need to be attribute values, rather than comments.

What about XML? It's one of your original suggestions, can be easily validated, has very widespread support, and per _"I think comments would be useful"_, it supports comments.

> I must be missing something. You currently have a single file with everything jammed into it (setDraft03, setDraft04, etc). The same piece of code knows about all the different specs, which means it will grow ad infinitum until the end of time.

I misunderstood you then sorry. The same *file* currently contains all the spec versions, but they don't know about each other, and do not extend each other. They each extend a single common base (i.e. the class constants). I agree with you that changing the architecture of how these are organised is desirable; you've made some very good points and I like the way you're thinking.

**Comment by erayd** [comment]
Would you like the spec reference comments on the class constants split up somehow?

**Comment by shmax** [comment]
Fine, no YAML. 

Now that I think about it a bit, encoding the rules as JSON is probably not a terrible idea, because as you say, the "comments" as attributes could be of use themselves to consuming libraries.

>  but they don't know about each other, and do not extend each other

Buddy, if we're going to get through this you're going to have to stop arguing semantics with me over every little thing. When I say they know about each other, I mean they're in the same code, which is true. They know about each other in the sense that all governed in the same file by the same parent--if you make a mistake editing one of the `set<draft>` functions you risk messing up the others--and in that's enough. It sounds like you agree that they should be separated, and that's good and all you had to say.

**Comment by erayd** [comment]
> When I say they know about each other, I mean they're in the same code, which is true. 

Works for me :-).

I care about semantics, because I've been burned way too many times by people making assumptions about what a particular wording means (especially as regards contract scoping) - but all that really matters is that we are on the same page. I'll try to tone down the nitpicking some, and please feel free to call me out if you feel I'm doing so needlessly (and you're right that picking at the 'knows about' point was probably needless, as we're splitting it *anyway*).

However, if I think we might be talking about different things, and it matters that we *not* be doing that, it does seem reasonable to clarify meaning. I will try to clarify or request clarification for only those situations that matter, and leave the rest alone. Does that sound like a reasonable stance to take?

**Comment by shmax** [comment]
Sure, all is happy and good. I'll try to quit being an old grouch (no promises). 

So what do we do next?

**Comment by erayd** [comment]
I think deciding how we want to organise the rule documents is a good start. Once we have that, then we can move on to the PHP side of things that interfaces with them.

My current thoughts are:
 * A document that defines all the rules and their descriptions, but does not give them values.
 * A schema that validates ^ for the purposes of unit testing.
 * For each draft, a document that gives every rule a value, and a reference to the spec.
 * A schema that validates ^ for the purposes of unit testing.

This gives us a clear framework, and means that we can check:
 * That all rules are deliberately given a value for each draft, rather than being accidentally inherited
 * That all rules are actually present for each draft
 * That a draft doesn't define any extra rules that aren't in the base document

It also means that the spec reference comments won't grow over time, because they'll be in the draft documents, rather than the base.

A modification to this approach could be to define defaults in the base rule file, which would allow much more concise files for each draft, but has the downside of requiring the spec references to all be crammed in the same place again, and increases the chance of accidentally inheriting incorrect default values.

Thoughts?

**Comment by shmax** [comment]
Sorry, meeting (re: ping-pong). Need to do a little work, but will come back to this in a bit. So far everything you say SGTM.

**Comment by erayd** [comment]
All good - no huge rush :-)

**Comment by shmax** [comment]
Do you think we should submit these rules documents to the main spec repo? I feel that's where it should live, but on the other hand they seem to bicker a lot and it might be best for us to work with it somewhere independent for a while to work the kinks out first...

**Comment by erayd** [comment]
I'd prefer to keep it away from there for now, because if we try to sort it out over there it will likely take months, and I need this sooner than that. The bickering is a big factor, plus the much larger number of strong and conflicting opinions, plus after reading through some of the issue history I have concerns about the overall project management over there.

I like working with you, because we usually have some robust discussion and end up with a better overall result, but it never feels like work has stalled, and we tend to reach agreement in a reasonably timely fashion for most things, even if our initial opinions differ significantly.

Long term having the definitions in the spec repo would certainly be nice, but I'd prefer to have them in a robust and usable state before we go down that road.

**Comment by shmax** [comment]
Right on. I figured, just thought I'd check.

**Comment by erayd** [comment]
Good call :-)

**Comment by erayd** [comment]
What do you think of [this](https://github.com/erayd/json-schema-info/tree/rules/rules) for the rule file formats?

I've just done the type rules, so you can see what I'm going for - if you like it, then I can add the rest :-).

**Comment by shmax** [comment]
So far so good!

**Comment by erayd** [comment]
Have tweaked a little to remove some redundancy and improve specificity of errors and ease-of-use when referencing rules. Are you happy with the approach I've taken, or would you like changes?

**Comment by erayd** [comment]
What do you think of [this](https://github.com/erayd/json-schema-info/tree/common-base)?

I couldn't think of any sane way to do the split and still keep the references, and after a lot of thought I decided that splitting was probably more important than keeping the references. So I got rid of them.

**Comment by shmax** [comment]
Looking really sharp, but I might need a little tour of the repo; where are the actual rules for each draft encoded? I dug around but all I could find were what looked like skeletal files...

**Comment by erayd** [comment]
The rules for each draft are in rules/standard/*.json.

**Comment by shmax** [comment]
You mean this? https://github.com/erayd/json-schema-info/blob/common-base/rules/standard/draft-04.json

I must be dense--still don't get it. Weren't there going to be a bunch of booleans and whatnot? This looks more like a set of keywords...

**Comment by erayd** [comment]
Yep, that's exactly what I mean.

For each rule in that file, it merges whatever definition is there with the corresponding rule data in `rules/base.json`, and returns a `RuleInfo` object. If you ask for a rule which is *not* defined, then it returns null.

`RuleInfo` will lazy-load undefined informational attributes from rules/schema.json as they are requested.

**Comment by erayd** [comment]
See usage examples [here](https://github.com/erayd/json-schema-info/blob/common-base/tests/SchemaInfoTest.php#L102).

**Comment by shmax** [comment]
I'm afraid I didn't glean much from the unit test code you linked to. Let's say I'm writing a validator, and I want to know if the `required` schema property should be a boolean value or an array of strings. What would that look like?

**Comment by erayd** [comment]
```PHP
$info = new SchemaInfo('http://json-schema.org/draft-04/schema#');

if ($info->standard('requiredIsBoolean')) {
  // required is a boolean
} else {
  // required is not a boolean
}
```

**Comment by shmax** [comment]
Gotcha. So when you say it lazy-loads attributes, does that mean it does a disk read for each new attribute it has to query?

**Comment by erayd** [comment]
No - I mean [this](https://github.com/erayd/json-schema-info/blob/common-base/src/RuleInfo.php#L39). Most uses don't actually need the info stuff, so I figured no sense in wasting resources on them unless the user actually wants them. I'm probably over-optimising a bit.

**Comment by shmax** [comment]
Sorry to keep you hanging--I'm distracted by some other projects right now, but will try to give you my full attention this weekend; I'll grab your branch, try it out, understand how it works, provide better feedback, the whole deal.

**Comment by erayd** [comment]
No problem, and thanks - I really appreciate that :-).

Let me know if you have any questions about it, or if you'd like examples. I'll put up the latest round of dogfood once I've finished making it work with draft-06 so you can see how it looks when integrated into a validator.

**Comment by shmax** [comment]
Started taking a look at this, and I have some initial feedback:

1. Awesome work!
2. Still not a fan of your idea of taking a url input and regex'ing the bit you're interested in out of it and discarding the rest. It's misleading to the user, who would be justified in thinking that he could insert his own arbitrary url, and then you're going to get issues along the lines of "why doesn't my url get hit?". If you're only interested in the "draft-0x" bit, then that's what the input should be.
3. I spent about 20 minutes being utterly mystified by this line (from your sample):

    ```PHP
    if ($info->standard('requiredIsBoolean')) {
    ```
    Granted, I had my own expectations working against me, but I intuited that a boolean would be coming out of that check (yes, I know you explained it earlier), and when I stepped through in the debugger I couldn't find one anywhere. I then searched the codebase for "requiredIsBoolean", and found that it was only in draft-03, with a value of `{}`. I know we've already been over this ground, but it just feels counter-intuitive to me. 

    When I suggested splitting your rules apart into a base and overrides, what I had in mind was something more like this:

    ```js
    // ...,
    {
       "allowUndefinedTypes":true, // "Whether other, non-spec primitive types are allowed"
       "selfDescriptiveMetaSchema": false, // "Whether the meta-schema must be valid against itself"
       // ...
       "requiredIsBoolean":true // "Whether 'required' is a boolean"
    }
    ````
    The default for each one would be whatever the value would be for the earliest draft, and we demote the `description` property to a comment, as it didn't seem to have any run-time purpose, anyway (if you do want to keep it, you could let it stay as-is and add a 'value' bucket for each rule)

    Then, any particular draft that has a rule that differs from the default overrides it, like this:
    ```js
    // draft-04.json
    {
        // ...
        'standard': {
            // ...
           'requiredIsBoolean' : false
        }
    }
    ````

    Finally, in your SchemaInfo constructor, you just do something like this:
    ````php
        $this->specInfo = utils::extend($baseInfo, $specInfo); (I can provide the extend method I use if you need it)
    ````

Now you're left with an array that can be nested to any depth with all the information in it in an intuitive format (booleans, and such). I figure all the `RuleSetInfo` stuff can go away (or at least be deployed as a wrapper around the final result); it doesn't really seem to do much and you'll probably want as little overhead as possible.

Anyway, that's my two cents. I can provide a PR if you're interested.

**Comment by erayd** [comment]
Thanks heaps for your feedback :-).

> Still not a fan of your idea of taking a url input and regex'ing the bit you're interested in out of it and discarding the rest.

I did this so that you can just pass the constructor a spec URI. If you feed it an arbitrary other URI, then it'll throw an exception - there is no ambiguity about what is valid. I used a regex rather than a switch / case block in order to catch URIs with http, and URIs without a fragment, or with an arbitrary fragment.

I feel this behavior is important - if the user passes a URI that looks like a spec URI to the constructor, they should get a corresponding object. Can you clarify which part of this is confusing to the user? I think I might be missing something here - to me, the URI behavior appears obvious, but I wrote the thing, so I may be too close to see the problem.

> When I suggested splitting your rules apart into a base and overrides, what I had in mind was something more like this...default for each one would be whatever the value would be for the earliest draft...

Just to clarify - your suggestion would mean dragging stuff from draft-03 around for all eternity within the other rulesets. Is that really what you are intending? For example, all the discarded formats and keywords that are no longer valid after draft-03 would need to be present in draft-04, draft-06, and all future drafts, just so that they can be explicitly overridden to false.

I don't suppose you'd be happy to have a discussion via some form of IM (IRC maybe?) so that we can get on the same page? We seem to be going back and forth a bit on the fundamental architecture, and I get the feeling we have slightly different interpretations of a few things here, so that seems like the fastest way to proceed if you'd be OK with it :-).

**Comment by shmax** [comment]
> Can you clarify which part of this is confusing to the user?
If I see a method that takes a url as an argument, my expectation is that it is going to send a request to that url. In your sample, you are passing in a url to a real location. I would be absolutely justified in thinking that an http request is going to happen. I would be further justified in thinking that I could, say, host my own schema file somewhere with a few modifications and use that. Only you and I know that the black box behavior is to do actually do some regex fishing and do no request whatsoever.

If you don't intend to use a url, then why invite one? You already have your `SPEC_DRAFT_0x` constants; why not use one of them?

> Just to clarify - your suggestion would mean dragging stuff from draft-03 around for all eternity within the other rulesets. Is that really what you are intending? For example, all the discarded formats and keywords that are no longer valid after draft-03 would need to be present in draft-04, draft-06, and all future drafts, just so that they can be explicitly overridden to false.

Well, in base.json--not repeated in the draft files. But their presence was really only for clarity; if you think runaway growth is really a concern you could just not do the extend and consider anything not found to be falsy. You would have to adopt the convention that all rules are named in the affirmative, but that's probably good practice, anyway.

**Comment by shmax** [comment]
Oh, and sure. I'll send you my IM info...

**Comment by erayd** [comment]
Not seeing anything from you with IM info - I'll keep an eye on my inbox, and in the meantime I'm sitting in #json-schema-info on freenode IRC.

**Comment by shmax** [comment]
Had to reset my password, which is always an adventure with AIM. Patience...

**Comment by erayd** [comment]
How's [this](https://github.com/erayd/json-schema-info/tree/common-base)?

**Comment by shmax** [comment]
I'm diggin' it... very cool 🥇

**Comment by erayd** [comment]
Do I rustle up a PR for it?

**Comment by shmax** [comment]
Of course!

**Comment by erayd** [comment]
Yay! Stand by...

**Comment by erayd** [comment]
Implemented in #5.
